crates/ra_hir_ty/src/infer/expr.rs

```diff
@@ -63,6 +63,75 @@ impl<'a> InferenceContext<'a> {
         self.resolve_ty_as_possible(ty)
     }
 
+    fn callable_sig_from_fn_trait(&mut self, ty: &Ty) -> Option<FnSig> {
+        if let Some(krate) = self.resolver.krate() {
+            let fn_traits: Vec<crate::TraitId> = [FnTrait::FnOnce, FnTrait::FnMut, FnTrait::Fn]
+                .iter()
+                .filter_map(|f| get_fn_trait(self.db, krate, *f))
+                .collect();
+            for fn_trait in fn_traits {
+                let fn_trait_data = self.db.trait_data(fn_trait);
+                let generic_params = generics(self.db.upcast(), fn_trait.into());
+                if generic_params.len() != 2 {
+                    continue;
+                }
+
+                let arg_ty = self.table.new_type_var();
+                let substs = Substs::build_for_generics(&generic_params)
+                    .push(ty.clone())
+                    .push(arg_ty.clone())
+                    .build();
+
+                let trait_ref = TraitRef { trait_: fn_trait, substs: substs.clone() };
+                let trait_env = Arc::clone(&self.trait_env);
+                let implements_fn_goal =
+                    self.canonicalizer().canonicalize_obligation(InEnvironment {
+                        value: Obligation::Trait(trait_ref),
+                        environment: trait_env,
+                    });
+                if let Some(Solution::Unique(SolutionVariables(solution))) =
```

Comment:
2. makes sense, but don't we still need to destruct the solution to get the argument types out? For example, if the solution is `Ambig(Guidance::Unknown)` then we won't be able to create the vector of argument types.

Reply:
Not if we make the query with a tuple already. I.e. Instead of creating one variable and asking for `Fn<?1>`, make one variable per parameter and ask for `Fn<(?1, ?2, ?3, ...)>`. Then we can just return those variables. (By registering the obligation, the solution will be automatically applied to those variables once it is found. We *could* of course already use the solution here if it is unique, but that'd only be an optimization.)